### ISSUE TITLE
Parallel threads

### ISSUE THREAD

**Issue by paolafer** [next-exp/IC#528]
I'm not sure if it's an IC problem, just tell me if it is not.
I've been running diomira in the majorana1 queue system and I noticed that it was extremely slow, much more (> 10 times) than if I ran in the same machine interactively.
I've asked José Vicente about it and he says that each process is being parallelized to 44 threads and this is the reason why everything is so slow. I don't have any number of threads defined in my scripts for the queue system (and I haven't changed them lately), so I was wondering if there have been any changes in the IC code that could affect the thread parallelization. I'm totally ignorant about this topic, so if anyone can help me, I'd be grateful.

**Comment by jjgomezcadenas** [comment]
Diomira certainly is nor prepared for paralelization, but I see no reason why it should run so much slower, we need to talk with JV about this.

**Comment by paolafer** [comment]
José Vicente suggested to set the environment variable OMP_NUM_THREADS to 1, explicitly:

`export OMP_NUM_THREADS=1`

and it worked. However, I wonder why we didn't need to do anything before.

**Comment by paolafer** [comment]
The issue is closed. Please, take note of the solution, in case you need it at some point.

**Comment by mmkekic** [comment]
I reopened this issue because it is not solved and might be handy to look at it from time to time :)

**Comment by gonzaponte** [comment]
So @mmkekic and I have been looking into the guts of this issue. This is what we've learned:

`numpy` and other modules use some parallelization libraries internally to speed things up. This is a design choice that is a bit obscure and barely documented. However, it is also a configurable behavior. There are a number of environmental variables that set the maximum number of threads to be used by each of those parallelization libraries:

```
export MKL_NUM_THREADS=1
export NUMEXPR_NUM_THREADS=1
export OMP_NUM_THREADS=1
```

But the story doesn't end there. After setting those variables to 1 (so each run of the city command is single-threaded) there was something still running in two threads. I tracked down it to a `np.cos` call in `sierpe/low_frequency_noise.py`. [Apparently](https://stackoverflow.com/questions/59445147/weird-bug-in-pandas-and-numpy-regarding-multithreading), `numpy` and `pandas` have different design choices regarding the number of threads to use. While `pandas` uses internally `numexpr`, which sets the maximum number of threads to 1 when imported, `numpy` does something different (not sure what or how) and has some basic functions parallelized (`np.sqrt`, `np.cos`, etc.). In principle, setting the number of threads to 1 would solve this issue but, after trying to do so in the `city` command, nothing changed. This means there are some complex interactions between imports and I-don't-know-what that translates into things still being parallelized.

So to stop the parallelization in the `np.cos` call I've added a `numexpr.set_vml_num_threads(1)` right before that line and it worked. This doesn't seem to be necessary when running on the computer farm, according to @mmkekic, but maybe we can explore the possibility of adding it for everybody provided there is no performance penalty.

Finally, while researching this I've found some other variables that are not affecting us at the moment, but that may be relevant if this problem arises in the future, so I will list them here:

```
export OPENBLAS_NUM_THREADS=1
export VECLIB_MAXIMUM_THREADS=1
```

**Comment by gonzaponte** [comment]
We want to take advantage of this small modification to the code to introduce new people to PRs.
Whoever wants to take this should
1. Make sure the local and origin master branch is at the same commit as nextic/master
2. Create a new branch starting on master. Checkout to this new branch.
3. Make a dummy commit (add a new line or an empty space) and push to origin.
4. Go to the github page and open a new PR from your branch to nextic/master

There, we will discuss the possibilities. I see mainly three:
1. Set these variables in the city command
2. Add a command line option to set the variables
3. Add a new function to manage.sh setting these variables

With option 1 the variables are set only at city run time, not affecting other things (such as IC clients). However, this means that the city cannot be run on the client's computer taking advantage of parallelization.
Option 2 would require to add a command line option in `configure.py` and doing the job in the `city` function (not the command). This way, for jobs running in the cluster, one would be able to turn off the parallelization without the regular user having to deal with it explicitly.
Option 3 also means that the regular user doesn't have to do anything explicitly (only when running in the cluster), but after setting these variables I am not sure resetting them would work.

Right now I would go for option 2. It seems the most robust, but it may be also the most technical one and, naturally, it's open for discussion.

**Comment by jacg** [comment]
You should also make sure that you are aware of [numexpr](https://github.com/pydata/numexpr) and what it can offer you in this sort of area.

**Comment by gonzaponte** [comment]
May I ask you to elaborate a bit on that? Do you mean that we should consider using `numexpr` directly in our code?

**Comment by jacg** [comment]
Sorry, after reading your message more carefully, I now see that you already explicitly mention numexpr ... I merely saw you talking about controlling numbers of threads, and thought that you should know that numexpr does this sort of stuff for you ... so while I thought I was pointing out a new and useful tool to you, it seems that you're already aware of numexpr (to some extent) and struggling *against* it.

So, in case you are not aware, numexpr does (or at least used to do) 3 interesting and useful things:

1. Avoid temporaries in numpy expressions. For example `a * b + c` first calculates and allocates `a * b` which is then used to calculate and allocate the final answer. The result of `a * b` is a temporary which disappears when the final answer is produced. Allocating these temporaries is wasteful. Numexpr only allocates the final result.

2. Optimize cache usage. If the matrices involved in a numpy expression don't fit in cache, the CPU spends a lot of time waiting around for data to be read into cache. Numexpr breaks the calculation down so that it works on smaller blocks within the matrices, small enough to fit in cache, in order to minimize cache invalidation.

3. Go behind the GIL and take advantage of any cores you might have lying around.

So, yes, I do think that you should consider using numexpr directly in your code. I don't know whether your specific code can benefit from it, bit in cases where you have complex numpy expressions involving large matrices and some spare cores, numexpr can give some impressive speedups.

**Comment by gonzaponte** [comment]
Thanks for the summary. I've been reading about `numexpr` while tackling this particular issue and the problem we have is precisely that we cannot take advantage of `numexpr` trickeries. This is because we "*have to*" run on a single thread. When running on our computer cluster, each job is restricted to a single processor. This means that any program that tries to parallelize a code will simply use different threads within the same processor, which makes it (noticeably) slower. That's the only reason we are trying to restrict the whole program to a single thread.

Now, I am certain that this is a virtual problem, meaning that the cluster should be able to handle jobs that go into many processors and take advantage of the parallelization. I've seen the same software we use being used in that way, but I am far from understanding how it is managed behind the scenes. This is maybe something we can try to address with the people that handle the computer cluster, before trying to mess with the number of threads in our own code.

If this turns out to be a virtual problem, we will take into account `numexpr` in the future (thanks for the advice). However, most of our code is not particularly slow and the effort that would have to be invested in switching from `numpy` to `numexpr` is probably too high to be worth it. On the other hand, we have a new *city* under review that may take advantage of it, so it may come in handy!

**Comment by jacg** [comment]
> we "have to" run on a single thread

OK, gotcha.

> switching from `numpy` to `numexpr`

You make it sound more complicated than it usually is: you take an numpy expression, feed it to numexpr, and it now runs faster.

In that sense it's a bit like numba, it's not that you 'switch from python to numba' you feed python to numba and it runs faster.

**Comment by gonzaponte** [comment]
> You make it sound more complicated than it usually is: you take an numpy expression, feed it to numexpr, and it now runs faster.

That's right, we should consider applying numexpr to the bottlenecks of our code. Thanks!
